Feature/drop value modifiers

hussain-jafari opened the pull request:
## Drop value modifiers

### Description
- *Category*: implementation
- *JIRA issue*: [MIC-3904](https://jira.ihme.washington.edu/browse/MIC-3904)
- *Research reference*: [Affected Outcomes](https://vivarium-research.readthedocs.io/en/latest/models/intervention_models/crm_mgmt/vivarium_lifestyle.html#affected-outcomes)

### Changes and notes
Added BMI with drop value exposure pipeline because value modifier population view needs access to columns which aren't present at that stage in initialization. Code for initializing last FPG test date shows up in diff after resolving merge conflicts.  

### Verification and Testing
Tested that enrolled simulants had their drop value pipelines updated, and would remain in maintenance period for the appropriate amount of time. Verified that there was a linear decrease in drop value over the specified period and stabilized to the correct final drop value.

hussain-jafari committed:
intermediate push

hussain-jafari committed:
value modifiers (without adherence)

hussain-jafari committed:
store bmi without drop value pipeline name

hussain-jafari committed:
black and isort on risks and data_values

hussain-jafari committed:
black and isort on treatment

hussain-jafari committed:
lifestyle adherence (#107)

    Category: implementation
    JIRA issue: MIC-3905
    Research reference: lifestyle dropout

Verification and Testing
Looked at lifestyle adherence column in an interactive context and checked that ratio of adherent simulants was roughly what we wanted.

hussain-jafari committed:
include adherence

hussain-jafari committed:
Merge branch 'main' into feature/drop_value_modifiers

rmudambi commented on src/vivarium_nih_us_cvd/components/risks.py:
```diff
@@ -26,6 +26,9 @@ class DropValueRisk(Risk):
     def __init__(self, risk: str):
         super().__init__(risk)
+        self.exposure_without_drop_value_pipeline_name = (
+            f"{self.risk.name}_without_drop_value.exposure"
```
Our (unofficial) convention has been to call things like this `f"{self.risk.name}.raw_exposure"`

rmudambi commented on src/vivarium_nih_us_cvd/components/risks.py:
```diff
@@ -45,6 +51,14 @@ def _get_drop_value_pipeline(self, builder: Builder) -> Pipeline:
             source=lambda index: pd.Series(0, index=index),
         )
 
+    def _get_exposure_without_drop_value_pipeline(self, builder: Builder) -> Pipeline:
```
Following the convention above this could become `_get_raw_exposure_pipeline`

rmudambi commented on src/vivarium_nih_us_cvd/components/treatment.py:
```diff
@@ -240,6 +246,121 @@ def _register_target_modifiers(self, builder: Builder) -> None:
+        # drop value modifiers
+        builder.value.register_value_modifier(
+            data_values.PIPELINES.BMI_DROP_VALUE,
+            modifier=self._apply_lifestyle_to_bmi,
```
Consider doing something with partial functions like this so you don't have to define 3 functions that are almost identical:
```
def _apply_lifestyle(self, index: pd.Index, target: pd.Series, risk: str):
    # allow for updating drop value of dead people - makes interacting with target easier
    pop = self.population_view.get(index)
    enrollment_dates = pop[data_values.COLUMNS.LIFESTYLE]
    updated_drop_values = self.get_updated_drop_values(
        target, enrollment_dates, risk=risk
    )
    
return updated_drop_values
```

And when registering the value modifier you can do this:
```
builder.value.register_value_modifier(
    dava_values.PIPELINES.BMI_DROP_VALUE,
    modifier=partial(self._apply_lifestyle, "bmi"),
   ...
)
```

hussain-jafari replied:
Unfortunately this returns the following error

```
File ~/repos/vivarium_nih_us_cvd/src/vivarium_nih_us_cvd/components/treatment.py:250, in Treatment._register_target_modifiers(self, builder)
    240 builder.value.register_value_modifier(
    241     data_values.PIPELINES.LDLC_EXPOSURE,
    242     modifier=self.ldlc_target_modifier,
   (...)
    246     ],
    247 )
    249 # drop value modifiers
--> 250 builder.value.register_value_modifier(
    251     data_values.PIPELINES.BMI_DROP_VALUE,
    252     modifier=partial(self._apply_lifestyle, risk="bmi"),
    253     requires_columns=[
    254         data_values.COLUMNS.LIFESTYLE,
    255     ],
    256 )
    258 builder.value.register_value_modifier(
    259     data_values.PIPELINES.FPG_DROP_VALUE,
    260     modifier=partial(self._apply_lifestyle, risk="fpg"),
   (...)
    263     ],
    264 )
    266 builder.value.register_value_modifier(
    267     data_values.PIPELINES.SBP_DROP_VALUE,
    268     modifier=partial(self._apply_lifestyle, risk="sbp"),
   (...)
    271     ],
    272 )

File ~/repos/vivarium/src/vivarium/framework/values.py:626, in ValuesInterface.register_value_modifier(self, value_name, modifier, requires_columns, requires_values, requires_streams)
    592 def register_value_modifier(
    593     self,
    594     value_name: str,
   (...)
    598     requires_streams: List[str] = (),
    599 ):
    600     """Marks a ``Callable`` as the modifier of a named value.
    601 
    602     Parameters
   (...)
    624 
    625     """
--> 626     self._manager.register_value_modifier(
    627         value_name, modifier, requires_columns, requires_values, requires_streams
    628     )

File ~/repos/vivarium/src/vivarium/framework/lifecycle.py:384, in ConstraintMaker.constrain_normal_method.<locals>._wrapped(*args, **kwargs)
    380 self.check_valid_state(method, permitted_states)
    381 # Call the __func__ because we're rebinding _wrapped to the method
    382 # name on the object.  If we called method directly, we'd get
    383 # two copies of self.
--> 384 return method.__func__(*args, **kwargs)

File ~/repos/vivarium/src/vivarium/framework/values.py:386, in ValuesManager.register_value_modifier(self, value_name, modifier, requires_columns, requires_values, requires_streams)
    352 def register_value_modifier(
    353     self,
    354     value_name: str,
   (...)
    358     requires_streams: List[str] = (),
    359 ):
    360     """Marks a ``Callable`` as the modifier of a named value.
    361 
    362     Parameters
   (...)
    384 
    385     """
--> 386     modifier_name = self._get_modifier_name(modifier)
    388     pipeline = self._pipelines[value_name]  # May create a pipeline
    389     pipeline.mutators.append(modifier)

File ~/repos/vivarium/src/vivarium/framework/values.py:447, in ValuesManager._get_modifier_name(modifier)
    445     modifier_name = modifier.__name__
    446 elif hasattr(modifier, "__call__"):  # Some anonymous callable
--> 447     modifier_name = f"{modifier.__class__.name__}.__call__"
    448 else:  # I don't know what this is.
    449     raise ValueError(f"Unknown modifier type: {type(modifier)}")

AttributeError: type object 'functools.partial' has no attribute 'name__'
```

rmudambi replied:
In that case, retain the `self._apply_lifestyle` function and define three functions like 
```
def _apply_lifestyle_to_bmi(self, index: pd.Index, target: pd.Series):
    return self._apply_lifestyle(index, target, "bmi")
```

hussain-jafari committed:
add linearscaleup (#109)

    Category: implementation
    JIRA issue: MIC-3838
    Research reference: baseline coverage and scenarios

Verification and Testing
Looked at lifestyle exposure proportions across time steps by looking at non-baseline scenario, changing scale up start and end dates, and changing scale up start and end values.

hussain-jafari committed:
remove step size from scaleup logic

hussain-jafari committed:
Merge branch 'feature/drop_value_modifiers' of github.com:ihmeuw/vivarium_nih_us_cvd into feature/drop_value_modifiers

hussain-jafari committed:
address PR comments

hussain-jafari committed:
black and isort on treatment

rmudambi commented on src/vivarium_nih_us_cvd/model_specifications/nih_us_cvd.yaml:
```diff
@@ -78,7 +78,8 @@ configuration:
         location: 'Alabama'
-        artifact_path: '/ihme/costeffectiveness/artifacts/vivarium_nih_us_cvd/alabama.hdf'
+        #artifact_path: '/ihme/costeffectiveness/artifacts/vivarium_nih_us_cvd/alabama.hdf'
+        artifact_path: '/home/hussain-jafari/artifacts/alabama.hdf'
```
Fix the artifact path :)

rmudambi reviewed (approved):
Looks good.

hussain-jafari committed:
less duplicate code and fix artifact

hussain-jafari committed:
black again

hussain-jafari committed:
lifestyle scenarios (#110)

    Category: implementation
    JIRA issue: MIC-3819
    Research reference: baseline coverage and scenarios
